whipper/program/cdrdao.py

```diff
@@ -2,71 +2,172 @@
+                        'with CRC errors',
+                        (self.currentTrack, int(crc_s.group('channels'))))
+
+        self.oldline = line
+
+
+class ReadTOCTask(task.Task):
     """
-    Return cdrdao-generated table of contents for 'device'.
+    Task that reads the TOC of the disc using cdrdao
     """
-    # cdrdao MUST be passed a non-existing filename as its last argument
-    # to write the TOC to; it does not support writing to stdout or
-    # overwriting an existing file, nor does linux seem to support
-    # locking a non-existant file. Thus, this race-condition introducing
-    # hack is carried from morituri to whipper and will be removed when
-    # cdrdao is fixed.
-    fd, tocfile = tempfile.mkstemp(suffix=u'.cdrdao.read-toc.whipper')
-    os.close(fd)
-    os.unlink(tocfile)
-
-    cmd = [CDRDAO, 'read-toc'] + (['--fast-toc'] if fast_toc else []) + [
-        '--device', device, tocfile]
-    # PIPE is the closest to >/dev/null we can get
-    logger.debug("executing %r", cmd)
-    p = Popen(cmd, stdout=PIPE, stderr=PIPE)
-    _, stderr = p.communicate()
-    if p.returncode != 0:
-        msg = 'cdrdao read-toc failed: return code is non-zero: ' + \
-              str(p.returncode)
-        logger.critical(msg)
-        # Gracefully handle missing disc
-        if "ERROR: Unit not ready, giving up." in stderr:
-            raise EjectError(device, "no disc detected")
-        raise IOError(msg)
-
-    toc = TocFile(tocfile)
-    toc.parse()
-    if toc_path is not None:
-        t_comp = os.path.abspath(toc_path).split(os.sep)
-        t_dirn = os.sep.join(t_comp[:-1])
-        # If the output path doesn't exist, make it recursively
-        if not os.path.isdir(t_dirn):
-            os.makedirs(t_dirn)
-        t_dst = truncate_filename(os.path.join(t_dirn, t_comp[-1] + '.toc'))
-        shutil.copy(tocfile, os.path.join(t_dirn, t_dst))
-    os.unlink(tocfile)
-    return toc
+    description = 'Reading TOC'
+    toc = None
+
+    def __init__(self, device, fast_toc=False, toc_path=None):
+        """
+        Read the TOC for 'device'.
+
+        @param device:  block device to read TOC from
+        @type  device:  str
+        @param fast_toc:  If to use fast-toc cdrdao mode
+        @type  fast_toc: bool
+        @param toc_path: Where to save TOC if wanted.
+        @type  toc_path: str
+        """
+
+        self.device = device
+        self.fast_toc = fast_toc
+        self.toc_path = toc_path
+        self._buffer = ''  # accumulate characters
+        self._parser = ProgressParser()
+
+        self.fd, self.tocfile = tempfile.mkstemp(
+            suffix=u'.cdrdao.read-toc.whipper.task')
+
+    def start(self, runner):
+        task.Task.start(self, runner)
+        os.close(self.fd)
+        os.unlink(self.tocfile)
+
+        cmd = ([CDRDAO, 'read-toc'] + (['--fast-toc'] if self.fast_toc
+               else []) + ['--device', self.device, self.tocfile])
```

Comment:
Another stylistic change. I'd probably do something like
```python
        cmd = ([CDRDAO, 'read-toc']
               + (['--fast-toc'] if self.fast_toc else [])
               + ['--device', self.device, self.tocfile])
```
or even
```python
        cmd = (
            [CDRDAO, 'read-toc']
            + (['--fast-toc'] if self.fast_toc else [])
            + ['--device', self.device, self.tocfile]
        )
```
To make the sections easier to recognise. Putting a linebreak in the middle of a sub command seems ugly. (Breaking before binary operators as per [PEP 8: Should a Line Break Before or After a Binary Operator?](https://www.python.org/dev/peps/pep-0008/#should-a-line-break-before-or-after-a-binary-operator))